graph/ForceDirectedGraph/ForceDirectedGraph.tsx

```diff
@@ -0,0 +1,264 @@
+    DEFAULT_CLICKED_NODE_ID
+  ); // Node ID of clicked node
+  const [totalOffset, setTotalOffset] = useState({ x: 0, y: 0 });
+  const [scale, setScale] = useState(1);
+  const [lastScale, setLastScale] = useState(1); // Add state to keep track of last scale
+
+  const pressStartRef = useRef(0);
+
+  const screenWidth = Dimensions.get("window").width;
+  const screenHeight = Dimensions.get("window").height;
+
+  const panRef = useRef(null);
+  const pinchRef = useRef(null);
+
+  useEffect(() => {
+    const initialLinks = graph.getLinks();
+    const initialNodes = graph.getNodes();
+    const initialSizes = nodesSizesInit([...initialLinks]);
+    setLinks([...initialLinks]);
+
+    setNodes(
+      fruchtermanReingold(
+        [...initialNodes],
+        initialLinks,
+        constrainedNodeId,
+        screenWidth,
+        screenHeight,
+        MAX_ITERATIONS
+      )
+    );
+    setSizes(initialSizes);
+    setLoad(true);
+    return () => {
+      setLinks([]);
+      setNodes([]);
+      setSizes(new Map());
+      setLoad(false);
+    };
+  }, [graph, constrainedNodeId, screenWidth, screenHeight]);
+
+  // Update the draggable state of the whole graph when a node is clicked
+  useEffect(() => {
+    setDragEnabled(clickedNodeID === DEFAULT_CLICKED_NODE_ID);
+  }, [clickedNodeID]);
+
+  // Handle Dragging
+  const handlePanGestureEvent = (event: {
+    nativeEvent: { translationX: number; translationY: number };
+  }) => {
+    setTotalOffset({
+      x: event.nativeEvent.translationX / lastScale,
+      y: event.nativeEvent.translationY / lastScale,
+    });
+  };
+
+  // Handle Dragging State Change
+  const handlePanHandlerStateChange = (event: {
+    nativeEvent: { state: number };
+  }) => {
+    if (event.nativeEvent.state === State.END) {
+      setNodes(
+        nodes.map((node) => ({
+          ...node,
+          x: coordX(node),
+          y: coordY(node),
+        }))
+      );
+
+      if (!dragEnabled) {
+        setClickedNodeID(DEFAULT_CLICKED_NODE_ID);
+      }
+      setTotalOffset({ x: 0, y: 0 });
+    }
+  };
+
+  // Handle Zooming
+  const handlePinchGestureEvent = (event: {
+    nativeEvent: { scale: React.SetStateAction<number> };
+  }) => {
+    setScale(Number(event.nativeEvent.scale) * Number(lastScale));
+  };
+
+  const handlePinchHandlerStateChange = (event: {
+    nativeEvent: { state: number };
+  }) => {
+    if (event.nativeEvent.state === State.END) {
+      setLastScale(scale);
+    }
+  };
+
+  // Handle PRESS IN
+  const handlePressIn = (onPressCallback = () => {}) => {
+    onPressCallback();
+  };
+
+  // Handle PRESS OUT
+  const handlePressOut = (shortPressCallback = () => {}) => {
+    if (Date.now() - pressStartRef.current < SHORT_PRESS_DURATION) {
+      shortPressCallback();
+    }
+  };
+
+  // If the graph is not loaded, display a loading message
+  if (!load) {
+    return (
+      <View style={styles.container}>
+        <Text>Loading...</Text>
+      </View>
+    );
+  }
```

Comment:
For this, we should use an `ActivityIndicator` component. [ActivityIndicator](https://reactnative.dev/docs/activityindicator)